src/sql/src/pure.rs

```diff
@@ -106,34 +105,57 @@ pub async fn purify_create_source(
     PlanError,
 > {
     let CreateSourceStatement {
+        name,
         connection,
         format,
         envelope,
         include_metadata: _,
-        referenced_subsources: requested_subsources,
+        referenced_subsources,
         progress_subsource,
         ..
     } = &mut stmt;
 
-    if progress_subsource.is_some() {
-        bail_unsupported!("PROGRESS subsources not yet supported");
+    fn named_subsource_err(name: &Option<DeferredObjectName<Aug>>) -> Result<(), PlanError> {
+        match name {
+            Some(DeferredObjectName::Named(_)) => {
+                sql_bail!("Cannot manually ID qualify subsources")
+            }
+            _ => Ok(()),
+        }
     }
 
     // Disallow manually targetting subsources, this syntax is reserved for purification only
-    if let Some(ReferencedSubsources::Subset(subsources)) = requested_subsources {
+    named_subsource_err(progress_subsource)?;
+    if let Some(ReferencedSubsources::Subset(subsources)) = referenced_subsources {
         for CreateSourceSubsource {
             subsource,
             reference: _,
         } in subsources
         {
-            if let Some(DeferredObjectName::Named(_)) = subsource {
-                sql_bail!("Cannot manually ID qualify subsources");
-            }
+            named_subsource_err(subsource)?;
         }
     }
 
+    let mut subsource_id_counter = 0;
     let mut subsources = vec![];
 
+    let progress_desc = match &connection {
```

Comment:
feels like `connection` could have a function `get_progress_desc()` that encapsulates this match statement? But we can also do such a cleanup as a follow-up.

Reply:
I had the same thought but the parser is meant to be a crate only which traffics in parsing; we shouldn't hang semantic functions off of AST elements. I figured we could stick these things elsewhere if they're used commonly enough, but seemed like a simple enough concern that I decided to take the approach you suggest and do it in a follow up if it seems problematic enough.

Reply:
Great--appreciate the flexibility. Would be very glad to work on redesigning this if you have any ideas.